core/hotspot/cache/tinylfu.go

```diff
@@ -0,0 +1,182 @@
+package cache
+
+import (
+	"container/list"
+	"errors"
+)
+
+const (
+	samplesFactor            = 8
+	doorkeeperFactor         = 8
+	countersFactor           = 1
+	falsePositiveProbability = 0.1
+	lruRatio                 = 0.01
+)
+
+// TinyLfu is an implementation of the TinyLfu algorithm: https://arxiv.org/pdf/1512.00727.pdf
+//           Window Cache Victim .---------. Main Cache Victim
+//          .------------------->| TinyLFU |<-----------------.
+//          |                    `---------'                  |
+// .-------------------.              |    .------------------.
+// | Window Cache (1%) |              |    | Main Cache (99%) |
+// |      (LRU)        |              |    |      (SLRU)      |
+// `-------------------'              |    `------------------'
+//          ^                         |               ^
+//          |                         `---------------'
+//       new item                        Winner
+type TinyLfu struct {
+	countMinSketch *countMinSketch
+	doorkeeper     *doorkeeper
+	additions      int
+	samples        int
+	lru            *lru
```

Comment:
Maybe we need not to encapsulate window/probation/protected queue into lru and slru?
It might cause some confusion.